### PR TITLE
Nullable analysis of conditional operator should use best common type unless target typed

### LINKED ISSUE #72898: Compiler conversion crash with collectoin exprs.
Crash stack:

```
StreamJsonRpc.RemoteInvocationException: Object reference not set to an instance of an object.
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__154`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__18`1.MoveNext()
RPC server exception:
System.NullReferenceException: Object reference not set to an instance of an object.
      at Microsoft.CodeAnalysis.CSharp.ConversionsBase.ClassifyImplicitBuiltInConversionSlow(TypeSymbol source, TypeSymbol destination, CompoundUseSiteInfo`1& useSiteInfo)
      at Microsoft.CodeAnalysis.CSharp.ConversionsBase.ClassifyImplicitConversionFromType(TypeSymbol source, TypeSymbol destination, CompoundUseSiteInfo`1& useSiteInfo)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.GenerateConversion(Conversions conversions, BoundExpression sourceExpression, TypeSymbol sourceType, TypeSymbol destinationType, Boolean fromExplicitCast, Boolean extensionMethodThisArgument, Boolean isChecked)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConversion(BoundConversion conversionOpt, BoundExpression conversionOperand, Conversion conversion, TypeWithAnnotations targetTypeWithNullability, TypeWithState operandType, Boolean checkConversion, Boolean fromExplicitCast, Boolean useLegacyWarnings, AssignmentKind assignmentKind, ParameterSymbol parameterOpt, Boolean reportTopLevelWarnings, Boolean reportRemainingWarnings, Boolean extensionMethodThisArgument, Optional`1 stateForLambda, Boolean trackMembers, Location diagnosticLocation, ArrayBuilder`1 previousArgumentConversionResults)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.<VisitOptionalImplicitConversion>g__visitConversion|262_0(BoundExpression expr, TypeWithAnnotations targetTypeOpt, Boolean useLegacyWarnings, Boolean trackMembers, AssignmentKind assignmentKind, BoundExpression operand, Conversion conversion, TypeWithState operandType, Boolean delayCompletionForTargetType)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.<>c__DisplayClass262_0.<VisitOptionalImplicitConversion>b__2(TypeWithAnnotations targetTypeOpt)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.<>c__DisplayClass167_0.<VisitCollectionExpression>g__convertCollection|1(BoundCollectionExpression node, TypeWithAnnotations targetCollectionType, ArrayBuilder`1 completions)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.<>c__DisplayClass167_0.<VisitCollectionExpression>b__0(TypeWithAnnotations resultTypeWithAnnotations)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConversion(BoundConversion conversionOpt, BoundExpression conversionOperand, Conversion conversion, TypeWithAnnotations targetTypeWithNullability, TypeWithState operandType, Boolean checkConversion, Boolean fromExplicitCast, Boolean useLegacyWarnings, AssignmentKind assignmentKind, ParameterSymbol parameterOpt, Boolean reportTopLevelWarnings, Boolean reportRemainingWarnings, Boolean extensionMethodThisArgument, Optional`1 stateForLambda, Boolean trackMembers, Location diagnosticLocation, ArrayBuilder`1 previousArgumentConversionResults)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.ConvertConditionalOperandOrSwitchExpressionArmResult(BoundExpression node, BoundExpression operand, Conversion conversion, TypeWithAnnotations targetType, TypeWithState operandType, LocalState state, Boolean isReachable)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.<VisitConditionalOperatorCore>g__convertArms|209_0(BoundExpression node, BoundExpression originalConsequence, BoundExpression originalAlternative, LocalState consequenceState, LocalState alternativeState, TypeWithState consequenceRValue, TypeWithState alternativeRValue, BoundExpression consequence, Conversion consequenceConversion, Boolean consequenceEndReachable, BoundExpression alternative, Conversion alternativeConversion, Boolean alternativeEndReachable, TypeWithAnnotations resultTypeWithAnnotations, Boolean wasTargetTyped)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConditionalOperatorCore(BoundExpression node, Boolean isRef, BoundExpression condition, BoundExpression originalConsequence, BoundExpression originalAlternative)
      at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitConditionalOperator(BoundConditionalOperator node)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitExpressionWithoutStackGuard(BoundExpression node)
      at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
      at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitOptionalImplicitConversion(BoundExpression expr, TypeWithAnnotations targetTypeOpt, Boolean useLegacyWarnings, Boolean trackMembers, AssignmentKind assignmentKind, Boolean delayCompletionForTargetType)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitReturnStatement(BoundReturnStatement node)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitStatementsWithLocalFunctions(BoundBlock block)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitBlock(BoundBlock node)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.Visit(BoundNode node)
      at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitMethodBodies(BoundBlock blockBody, BoundBlock expressionBody)
      at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitNonConstructorMethodBody(BoundNonConstructorMethodBody node)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.Scan(Boolean& badRegion)
      at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Analyze(Boolean& badRegion, Optional`1 initialState)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.Analyze(NullableWalker walker, Symbol symbol, DiagnosticBag diagnostics, Optional`1 initialState, Builder snapshotBuilderOpt, Boolean requiresAnalysis)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.Analyze(CSharpCompilation compilation, Symbol symbol, BoundNode node, Binder binder, Conversions conversions, DiagnosticBag diagnostics, Boolean useConstructorExitWarnings, Boolean useDelegateInvokeParameterTypes, Boolean useDelegateInvokeReturnType, MethodSymbol delegateInvokeMethodOpt, VariableState initialState, MethodSymbol baseOrThisInitializer, Builder analyzedNullabilityMapOpt, Builder snapshotBuilderOpt, ArrayBuilder`1 returnTypesOpt, Boolean getFinalNullableState, VariableState& finalNullableState, Boolean requiresAnalysis)
      at Microsoft.CodeAnalysis.CSharp.NullableWalker.AnalyzeWithSemanticInfo(CSharpCompilation compilation, Symbol symbol, BoundNode node, Binder binder, VariableState initialState, DiagnosticBag diagnostics, Boolean createSnapshots, Boolean requiresAnalysis)
      at Microsoft.CodeAnalysis.CSharp.MethodCompiler.BindMethodBody(MethodSymbol method, TypeCompilationState compilationState, BindingDiagnosticBag diagnostics, Boolean includeInitializersInBody, BoundNode initializersBody, Boolean reportNullableDiagnostics, ImportChain& importChain, Boolean& originalBodyNested, Boolean& prependedDefaultValueTypeConstructorInitializer, InitialState& forSemanticModel)
      at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
      at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
      at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0()
      at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   --- End of stack trace from previous location ---
```

Repro: 

Using Version 17.10.0 Preview 2.0

Checkout roslyn main 8213da007023b0cd6e9c6b30d86e023daf82a592

open VS.  Go to file src\Analyzers\CSharp\Analyzers\UseCollectionInitializer\CSharpUpdateExpressionSyntaxHelper.cs

Change the following code in that file:

image

to:

image

Basically, replace `SpecializedCollections.SingletonEnumerable(` with `[`

### PR DESCRIPTION
**Pull request by cston**
Fixes #72898

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 1]
```diff
@@ -4339,23 +4339,18 @@ private BoundExpression BindValueConditionalOperator(ConditionalExpressionSyntax
...
                 hasErrors = true;
             }
             else
             {
                 trueExpr = GenerateConversionForAssignment(bestType, trueExpr, diagnostics);
                 falseExpr = GenerateConversionForAssignment(bestType, falseExpr, diagnostics);
                 hasErrors = trueExpr.HasAnyErrors || falseExpr.HasAnyErrors;
-                // If one of the conversions went wrong (e.g. return type of method group being converted
-                // didn't match), then we don't want to use bestType because it's not accurate.
-                type = hasErrors ? CreateErrorType() : bestType;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4358]: To fix the bug, it seems sufficient to either take this change (to use the best common type in binding always), or the change in NullableWalker (to use best common in nullable analysis always). But both changes seemed appropriate to me.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IFixedStatement.cs [hunk 2]
```diff
@@ -611,84 +610,76 @@ unsafe void M(bool b)
...
             Statements (1)
                 ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Int32*, IsInvalid, IsImplicit) (Syntax: 'p = b ? &i1 : &i2')
-                  Left: 
+                  Left:
                     ILocalReferenceOperation: p (IsDeclaration: True) (OperationKind.LocalReference, Type: System.Int32*, IsInvalid, IsImplicit) (Syntax: 'p = b ? &i1 : &i2')
-                  Right: 
-                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: ?, IsInvalid, IsImplicit) (Syntax: 'b ? &i1 : &i2')
-
+                  Right:
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: System.Int32*, IsInvalid, IsImplicit) (Syntax: 'b ? &i1 : &i2')
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IFixedStatement.cs:643]: It looks like this is the meaningful change in here. I didn't notice it until after looking at it for a bit.

Perhaps we should also exercise the difference in `semanticModel.GetTypeInfo(conditionalExprSyntax)` as a result of this change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4358]: Let me check my understanding.

1. In this path, bestType is not null and not ErrorType. If one of the conditional expr operands had an ErrorType, we would have inferred ErrorType as the bestType.
2. The removed lines here are checking if there is some error within the operands, which didn't cause them to have an ErrorType, per (1). We previously thought in this case that we need to change the conditional-expr type to ErrorType in this case. Now we are simply keeping the bestType which we already determined in the face of such errors.

Basically I'm not seeing what purpose the original check was serving, it feels like this change is just propagating along reasonable type information, which allows our assumptions about the bound tree shape to be met when we reach nullable analysis.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4358]: I do see that there is basically a conversion check from the operands to the best type, which can fail, and I also find that a bit confusing. Perhaps my issue is that we are not checking specifically if the conversion had errors, we are checking for any kind of errors within the whole operand.

Essentially if one of the operands doesn't convert to the "best type", why was that chosen as the "best type"? Are there some invariants in best type analysis we can rely on in this regard?

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4358]: `BestTypeInferrer.InferBestTypeForConditionalOperator()` and `GenerateConversionForAssignment()` are both using `ClassifyImplicitConversionFromExpression()`, so it's not clear when the `GenerateConversionForAssignment()` calls above would fail if the best common type is not an error type.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -5867,7 +5867,7 @@ void makeAndAdjustReceiverSlot(BoundExpression receiver)
 
             TypeSymbol? resultType;
             bool wasTargetTyped = node is BoundConditionalOperator { WasTargetTyped: true };
-            if (node.HasErrors || wasTargetTyped)
+            if (wasTargetTyped)
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5870]: I've added an assert rather than replacing the `if`.

**Review by RikkiGibson — APPROVED** [review]

**Comment by cston** [comment]
@dotnet/roslyn-compiler for a second review, thanks.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -37997,6 +37997,204 @@ static void Main()
...
+                Diagnostic(ErrorCode.ERR_SyntaxError, ")").WithArguments(",").WithLocation(7, 26),
+                // (7,27): error CS1003: Syntax error, ']' expected
+                //         return b ? x : [y);
+                Diagnostic(ErrorCode.ERR_SyntaxError, ";").WithArguments("]").WithLocation(7, 27));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var expr = tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Last();
+            Assert.Equal("y", expr.ToString());
+            _ = model.GetSymbolInfo(expr);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:38045]: Consider verifying the result since the change to `BoundConditionalOperator.NaturalTypeOpt` is surfaced through `.Type` from this API

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5) with a small test suggestion to consider
